File: go/binary_search/binary_search_big_count.go
package binary_search

// bigCount should return number of value occurences in a sorted array
func bigCount(arr interface {
	len() int
	get(int) int
}, val int) int {
	l, r := 0, arr.len()

	for l < r {
		m := (l + r) / 2

		if arr.get(m) < val {
			l = m + 1
		} else {
			r = m
		}
	}

	start := l

	l, r = 0, arr.len()

	for l < r {
		m := (l + r) / 2

		if arr.get(m) <= val {
			l = m + 1
		} else {
			r = m
		}
	}

	return r - start
}

File: java/src/binary_search/BinarySearchBigCount.java
package binary_search;

public class BinarySearchBigCount {
	public interface Array {
		long length();
		int get(long index);
	}

	// bigCount should return number of value occurences in a sorted array
	public static long bigCount(Array arr, int val) {
		long l = 0, r = arr.length();

		while (l < r) {
			long m = (l + r)/2;
			if (arr.get(m) >= val) {
				r = m;
			} else {
				l = m + 1;
			}
		}

		long l1 = l;
		r = arr.length();

		while (l < r) {
			long m = (l + r)/2;
			if (arr.get(m) > val) {
				r = m;
			} else {
				l = m + 1;
			}
		}

		return l - l1;
	}
}
